src/wp-includes/webfonts-api/class-wp-webfonts-registry.php

```diff
@@ -0,0 +1,241 @@
+<?php
+/**
+ * Webfonts API: Webfonts Registry
+ *
+ * @package WordPress
+ * @subpackage Webfonts
+ * @since 5.9.0
+ */
+
+/**
+ * Webfonts Registry.
+ *
+ * This registry exists to handle all webfonts.
+ *
+ * It handles the following within the API:
+ *  - loads the bundled provider files into memory;
+ *  - registers each provider with the API by:
+ *       1. creating an instance (object);
+ *       2. storing it in-memory (by its unique provider ID) for use with the API;
+ *
+ * @since 5.9.0
+ */
+class WP_Webfonts_Registry {
+
+	/**
+	 * An in-memory storage container that holds all registered webfonts
+	 * for use within the API.
+	 *
+	 * Keyed by font-family.font-style.font-weight:
+	 *
+	 *      @type string $key => @type array Webfont.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var array[]
+	 */
+	private $registered = array();
+
+	/**
+	 * Registration keys per provider.
+	 *
+	 * Provides a O(1) lookup when querying by provider.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var array[]
+	 */
+	private $registry_by_provider = array();
+
+	/**
+	 * Schema validator.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var WP_Webfonts_Schema_Validator
+	 */
+	private $validator;
+
+	/**
+	 * Creates the registry.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param WP_Webfonts_Schema_Validator $validator Instance of the validator.
+	 */
+	public function __construct( WP_Webfonts_Schema_Validator $validator ) {
+		$this->validator = $validator;
+	}
+
+	/**
+	 * Gets all registered webfonts.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return array[] Registered webfonts each keyed by font-family.font-style.font-weight.
+	 */
+	public function get_all_registered() {
+		return $this->registered;
+	}
+
+	/**
+	 * Gets the registered webfonts for the given provider.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param string $provider_id Provider ID to fetch.
+	 * @return array[] Registered webfonts.
+	 */
+	public function get_by_provider( $provider_id ) {
+		if ( ! isset( $this->registry_by_provider[ $provider_id ] ) ) {
+			return array();
+		}
+
+		$webfonts = array();
+		foreach ( $this->registry_by_provider[ $provider_id ] as $registration_key ) {
+			// Skip if not registered.
+			if ( ! isset( $this->registered[ $registration_key ] ) ) {
```

Comment:
Do we want to consider the case when a registration key is set to `null` if that's the case we might want to switch to `array_key_exists` instead.

Reply:
`! isset()` covers the case where `$registration_key` is `null`.

See [here](https://3v4l.org/18Ard).

Reply:
Oh I mean when the value on the array is `null` for instance:

```php
$keys = [
    'value' => null,
];


if ( isset( $keys[ 'value' ] ) {
   echo 'set';
} else {
    echo 'not set';
}
```

`Isset` would evaluate to `false` in that instance `array_key_exists` behave as expected instead.

Reply:
`isset()` is the right usage here. Why? The code within that conditional block is setting the element's value to `array()`. 

- If the key does exist but the element's value is `null`, this logic will change the value to `array()`. See it in action here https://3v4l.org/h9fAQ.
- If the key does not exist (meaning the element does not exist in the array), the element is created with its value initialized to an empty array.

Thank you @mitogh for raising this for further analysis and discussion. It's good to make sure the right things are being used in the right way. Cheers